Guled/redesign form and add retake and delete photo option

Somnibyte opened the pull request:
* Added `retake` and `delete` button menu and functionality. 
* `Comic Title` renamed to `Series Title`.
* `Series Title` and `Story Title` are now required fields.

Commit:
Redesign addCoverButton. Add retake and delete button. Add new assets.

Commit:
Display retake, delete, and divider when photo is taken.

Commit:
Add retake and delete button functionality. Fix formatting issues. Clean up code.

Add retake and delete button functionality.

Add resetCoverPhotoButton method.

Add alpha value to highlighted state of coverPhotoButton.

Fix logic associated with removing warning labels from textfields.

Add new alert method to allow for more alert options and custom actions. Refactor AddComicViewController code.

Configure textfield placeholders in xib.

Clarify comment. Make IBActions private.

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.350 (0.352) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 12.679 (12.681) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

mayasaxena commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -99,31 +104,27 @@ final class AddComicViewController: UIViewController {
+        // The attributed title for the coverPhotoButton is set on the storyboard, but if an attributed title is created for the highlighted state
+        // the text in the highlighted state will shift some reason. In order to avoid this, an attributed title is created for both
+        // the highlighted and normal state. The textAlignment and lineBreakMode are also set.
+        let coverPhotoButtonTitleNormal = NSAttributedString(string: "Take Cover Photo", attributes: [NSAttributedStringKey.font: LucienTheme.Fonts.muliBold(size: 16.0) ?? UIFont(), NSAttributedStringKey.foregroundColor: UIColor.white])
```
Pull all the cover photo button setup stuff out into a function called `setupCoverPhotoButton` and call that here.

mayasaxena replied:
Also looks like the image and text are a little off center?

Somnibyte replied:
I'll take a look into that centering issue right now.

mayasaxena commented on lucien-ios/Protocols/AlertDisplaying.swift:
```diff
@@ -19,4 +19,12 @@ extension AlertDisplaying where Self: UIViewController {
+        for action in actions {
+            alertController.addAction(action)
+        }
+        self.present(alertController, animated: true, completion: nil)
```
Remove `self`

Commit:
Make doneButtonTapped a private method. Move coverPhotoButton setup into a function.

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_1G71V6DSRF5XD-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.343 (0.345) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 13.298 (13.300) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

Commit:
Center cameraButtonIcon. Create BottomBorderTextField class. Refactor BottomBorderTextFieldClass.

Center coverPhotoButton image and text on highlight state. Create subclass of UITextField that creates borders at the bottom.

Refactor BottomBorderTextfield class.

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_1G71V6DSRF5XD-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.313 (0.314) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 11.343 (11.345) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

paulrolfe commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -178,19 +174,47 @@ final class AddComicViewController: UIViewController {
 
+    private func darkenTextFieldBottomBorder(textField: UITextField) {
+        guard let bottomBorderSubLayer = textField.layer.sublayers else { return }
+        let bottomBorder = bottomBorderSubLayer[0] as CALayer
```
I see this `bottomBorderSubLayer[0]` in several places. I see why that's convenient. But it's another case where I wouldn't want to rely on UI elements that you don't explicitly manage for this information. I would recommend you subclass UITextField and any textfields that have this bottom border can have a property called `bottomBorderLayer` that you can guarantee exists and is what you think it is.

Additionally if you use those subclassed UITextFields everywhere you can outsource some of the repetitive setup you're doing to that subclass.

paulrolfe commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -178,19 +174,47 @@ final class AddComicViewController: UIViewController {
     }
 
-    @IBAction func addCoverButtonTapped(_ sender: UIButton) {
+    private func showWarningLabel(label: UILabel, textField: UITextField, sublayers: [CALayer]) {
```
Why are the sublayers being passed in here? Don't they exist on the textfield. If you implement a subclass for these text fields this comment will be moot.

paulrolfe commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -136,13 +128,22 @@ final class AddComicViewController: UIViewController {
     }
 
+    private func setupCoverPhotoButton() {
+        // The attributed title for the coverPhotoButton is set on the storyboard, but if an attributed title is created for the highlighted state
```
I'd double check that this can't all be done in the xib/storyboard. Probably don't need attributed titles either, can just set the font and color separately.

paulrolfe replied:
Nevermind, that centeredAligned text is the issue, so you probably need the attributed strings. I'd just set it in the xib if you can though.

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_1G71V6DSRF5XD-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.361 (0.363) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 14.679 (14.681) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

paulrolfe commented on lucien-ios/Subclasses/BottomBorderTextField.swift:
```diff
@@ -0,0 +1,30 @@
+
+    /// Removes border of the UITextField and creates a border at the bottom of the textfield.
+    func createBottomBorder() {
+        let bottomBorder = CALayer()
```
A  way to improve this is to maintain a reference to this layer with an instance variable. Then when you update the color in the function below, you can just grab this property directly instead of checking the sublayers array for some layer which may or may not be the layer you wanted.

And you might be able to put this code into the initializers for the textfield so they don't have to be called by your view controller.

paulrolfe replied:
I feel like that comment was ambiguous, here's some throw away sample code.
```swift
final class BottomBorderTextField: UITextField {
    private let bottomBorder = CALayer()
    var borderColor: UIColor? {
        get {
            if let color = bottomBorder.backgroundColor {
                return UIColor(cgColor: color)
            } else {
                return nil
            }
        }
        set {
            bottomBorder.backgroundColor = newValue?.cgColor
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }

    func setup() {
        // add the layer and style it.
    }
}
```

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.458 (0.460) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 6.816 (6.818) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

paulrolfe commented on lucien-ios/AddNewComicBook/AddComicViewController.swift:
```diff
@@ -99,31 +104,17 @@ final class AddComicViewController: UIViewController {
-        configureTextFieldBorder(textField: releaseDateTextField)
-        releaseDateTextField.attributedPlaceholder = createPlaceHolderAttributedString(placeholder: "Year of Release")
+
+        seriesTitleTextField.createBottomBorder()
```
You don't need to call any of these `createBottomBorder` since that is done in the textfield initialization now.

paulrolfe commented on lucien-ios/Subclasses/BottomBorderTextField.swift:
```diff
@@ -0,0 +1,43 @@
+        createBottomBorder()
+    }
+
+    func createBottomBorder() {
```
this can be private, others don't need it since you're setting it up at initialization.

Commit:
Make createBottomBorder() private. Place release date textfield toolbar configuration code inside a method.

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_1G71V6DSRF5XD-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.361 (0.363) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 13.429 (13.431) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

ghost commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-13-FNIXXLMNILAIOKCFTRBFYTOIF6YFPO6UFMRN7E4V5TPRGRATODNQ/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_1G71V6DSRF5XD-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.339 (0.341) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 14.116 (14.118) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>
